subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/CacheBackedTaskHistoryRepository.java

```diff
@@ -379,12 +384,37 @@ private HistoricalTaskExecution loadPreviousExecution(TaskInternal task) {
         return ImmutableSortedSet.copyOf(outputPropertyNames);
     }
 
-    private static ImmutableSet<String> getDeclaredOutputFilePaths(TaskProperties taskProperties, StringInterner stringInterner) {
-        ImmutableSet.Builder<String> declaredOutputFilePaths = ImmutableSortedSet.naturalOrder();
-        for (File file : taskProperties.getOutputFiles()) {
-            declaredOutputFilePaths.add(stringInterner.intern(file.getAbsolutePath()));
-        }
+    private static ImmutableSet<String> getDeclaredOutputFilePaths(final TaskProperties taskProperties, final StringInterner stringInterner) {
+        final ImmutableSet.Builder<String> declaredOutputFilePaths = ImmutableSortedSet.naturalOrder();
+        FileCollectionInternal outputFiles = (FileCollectionInternal) taskProperties.getOutputFiles();
+        outputFiles.visitRootElements(new FileCollectionVisitor() {
+            @Override
+            public void visitCollection(FileCollectionInternal fileCollection) {
+                addAllPaths(fileCollection, declaredOutputFilePaths, stringInterner);
+            }
+
+            @Override
+            public void visitTree(FileTreeInternal fileTree) {
+                DeprecationLogger.nagUserOfDeprecated("Adding file trees which are not directory trees as output files", "Outputs should always be actual files");
```

Comment:
"Outputs should always be actual files" ... or directories? We just said that adding directory trees is okay?

Reply:
I changed the message.